### ISSUE TITLE
Update for latest CPython release

### ISSUE THREAD

**Issue by mattip** [pypy/pypy#4927]
CPython did a [security release](https://discuss.python.org/t/python-3-10-14-3-9-19-and-3-8-19-is-now-available)
We should
- [ ] Update libexpat to 2.6.0. This requires updating the build workers (linux Dockerfiles, windows dependency, macos via brew). Perhaps we should do like CPython and bundle libexpat into our sources for `pypy/modules/pyexpat` python/cpython#115474 which would statically link libexpat and reduce the number of bundled shared objects from 8 to 7
- [ ] Update the stdlib to get the fixes for zipfile, tempfile, urllib, socket test, hidden pth, iso2022_jp_3 test, libexpat test, etc
- [ ] Check that we properly handle the underlying socket function changed in https://github.com/python/cpython/pull/112600
- [ ] Check that we pass the iso2022_jp_3 test https://github.com/python/cpython/pull/111780
- [ ] Use X509_STORE_get1_objects instead of X509_STORE_get0_objects in _ssl https://github.com/python/cpython/pull/115548, raising better error in _ssl https://github.com/python/cpython/issues/107077
- [ ] Update bundled OpenSSL to 3.0.11
- [ ] Update zlib 1.3.1
- [ ] Use new libexpat API https://github.com/python/cpython/pull/116272

cc @mgorny

**Comment by cfbolz** [comment]
> Check that we pass the iso2022_jp_3 test https://github.com/python/cpython/pull/111780

we have vendored a (modified?) copy of `_codecs_iso2022.c`, so I think we need to do some fixes here.

**Comment by mattip** [comment]
I updated py3.9 to 3.9.19

**Comment by mattip** [comment]
Newly failing tests, to be investigated:
On py3.9
- [ ] lib-python.3.test.test_codecencodings_iso2022
- [ ] lib-python.3.test.test_multiprocessing_spawn
- [ ] lib-python.3.test.test_unicode

Additionally, on py3.10, maybe due to the way I updated the stdlib (only merged from py3.9, did not sync with cpython 3.10.14)
- [ ] lib-python.3.test.test_ensurepip
- [ ] lib-python.3.test.test_tarfile
- [ ] lib-python.3.test.test_tempfile
- [ ] lib-python.3.test.test_xml_etree
- [ ] lib-python.3.test.test_xml_etree_c

Additionally, should try to figure out where [this pytest collection failure](https://buildbot.pypy.org/summary/longrepr?testname=internal&builder=pypy-c-jit-linux-x86-64&build=9285&mod=pytest) is coming from, it appears as a line in the buildbot summary

[E](https://buildbot.pypy.org/summary/longrepr?testname=internal&builder=pypy-c-jit-linux-x86-64&build=9268&mod=pytest) [E](https://buildbot.pypy.org/summary/longrepr?testname=internal&builder=pypy-c-jit-linux-x86-64&build=9272&mod=pytest) [E](https://buildbot.pypy.org/summary/longrepr?testname=internal&builder=pypy-c-jit-linux-x86-64&build=9275&mod=pytest) [E](https://buildbot.pypy.org/summary/longrepr?testname=internal&builder=pypy-c-jit-linux-x86-64&build=9283&mod=pytest) [E](https://buildbot.pypy.org/summary/longrepr?testname=internal&builder=pypy-c-jit-linux-x86-64&build=9285&mod=pytest)  [pypy-c-jit-linux-x86-64](https://buildbot.pypy.org/summary?builder=pypy-c-jit-linux-x86-64)  pytest                                                         internal

**Comment by mattip** [comment]
The pytest error is coming from the -D tests

**Comment by mattip** [comment]
The tarfile failures come from `stat.filemode(stat.S_IMODE(36863))` which on CPython returns `'?rwsrwsrwt'` and on PyPy returns `'-rwsrwsrwt'`. The first letter is the file type.

**Comment by mattip** [comment]
We use the pure python implementation from lib-python/3/stat.py, CPython uses a c-implementation. The filemode implementation was fixed on CPython main in https://github.com/python/cpython/issues/108638. I resynced in 27da92029

**Comment by mgorny** [comment]
Thanks a lot!

**Comment by mgorny** [comment]
Any chance for a new release, by the way? I think we've collected quite some fixes already.

**Comment by mattip** [comment]
Yes, a release is a good idea. I want to stabilize the churn a bit first, there are some weird buildbot failures on windows. There are also a few milestone issues to resolve
